crates/c-api/src/store.rs

```diff
@@ -259,3 +266,119 @@ pub extern "C" fn wasmtime_context_set_epoch_deadline(
 ) {
     store.set_epoch_deadline(ticks_beyond_current);
 }
+
+#[cfg(unix)]
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_context_insert_file(
+    context: CStoreContextMut,
+    guest_fd: u32,
+    host_fd: RawFd, // int
+    access_mode: u32,
+) {
+    let access_mode = FileAccessMode::from_bits_truncate(access_mode);
+
+    // SAFETY: no other functions should call `from_raw_fd`, so there is only
+    // one owner for the file descriptor.
+    let f = unsafe { File::from_raw_fd(host_fd) };
+    let f = cap_std::fs::File::from_std(f);
+    let f = wasmtime_wasi::sync::file::File::from_cap_std(f);
+    context
+        .data()
+        .wasi
+        .as_ref()
+        .unwrap()
+        .insert_file(guest_fd, Box::new(f), access_mode);
+}
+
+#[cfg(unix)]
```

Comment:
Like with C, can the `#[cfg]` be used to conditionally define a type and delegate to a small helper internally? As I mentioned before that would avoid duplicating the whole function.

Reply:
I don't think so since we have different input type. Unless you want to eliminate that? I didn't find a good way to conditionally define ONLY the func header.